## Phase 1 : Bagging and Boosting
Concepts:
Ensemble learning, Bagging (Random Forest), Boosting (Adaboost, XGBoost)


In [ ]:
import pandas as pd
import numpy as np

Step 1:Data Understanding

In [ ]:
df = pd.read_csv('/content/breast-cancer.csv')
df.head()
df.info()
df.isnull().sum()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

Step 2: Preprocessing the Data

In [ ]:
df = df.drop(columns=['id'])    #Droped the 'id' column as it is a identifier
df['diagnosis'] = df['diagnosis'].map({'M': 1, 'B': 0})    #Encoded the 'diagnosis' column
X = df.drop('diagnosis', axis=1)
y = df['diagnosis']
print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (569, 30)
Target shape: (569,)


Step 3: Train-Test Split

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
print("Training set shape:", X_train.shape)
print("Test set shape:", X_test.shape)

Training set shape: (455, 30)
Test set shape: (114, 30)


Step 4: Model Training — RandomForestClassifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
rf_preds = rf_model.predict(X_test)
rf_acc = accuracy_score(y_test, rf_preds)
print("Random Forest Accuracy:", rf_acc)
print("\nClassification Report:\n", classification_report(y_test, rf_preds))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, rf_preds))


Random Forest Accuracy: 0.9736842105263158

Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98        72
           1       1.00      0.93      0.96        42

    accuracy                           0.97       114
   macro avg       0.98      0.96      0.97       114
weighted avg       0.97      0.97      0.97       114


Confusion Matrix:
 [[72  0]
 [ 3 39]]


Step 5: Train AdaBoostClassifier

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

ada_model = AdaBoostClassifier(random_state=42)
ada_model.fit(X_train, y_train)
ada_preds = ada_model.predict(X_test)
ada_acc = accuracy_score(y_test, ada_preds)
print("AdaBoost Accuracy:", ada_acc)
print("\nClassification Report:\n", classification_report(y_test, ada_preds))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, ada_preds))


AdaBoost Accuracy: 0.9824561403508771

Classification Report:
               precision    recall  f1-score   support

           0       0.97      1.00      0.99        72
           1       1.00      0.95      0.98        42

    accuracy                           0.98       114
   macro avg       0.99      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114


Confusion Matrix:
 [[72  0]
 [ 2 40]]


Step 6: Train XGBoostClassifier

In [ ]:
from xgboost import XGBClassifier
xgb_model = XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42)
xgb_model.fit(X_train, y_train)
xgb_preds = xgb_model.predict(X_test)
xgb_acc = accuracy_score(y_test, xgb_preds)
print("XGBoost Accuracy:", xgb_acc)
print("\nClassification Report:\n", classification_report(y_test, xgb_preds))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, xgb_preds))

XGBoost Accuracy: 0.956140350877193

Classification Report:
               precision    recall  f1-score   support

           0       0.94      1.00      0.97        72
           1       1.00      0.88      0.94        42

    accuracy                           0.96       114
   macro avg       0.97      0.94      0.95       114
weighted avg       0.96      0.96      0.96       114


Confusion Matrix:
 [[72  0]
 [ 5 37]]


/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [18:02:13] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


## Final Summary: Ensemble Learning on Breast Cancer Dataset

### Overview
 Applied and compared three ensemble learning models:
- **RandomForestClassifier**
- **XGBoostClassifier**

###  Model Performance Comparison

| Model         | Accuracy | Precision (M) | Recall (M) | F1-Score (M) | False Negatives |
|---------------|----------|----------------|--------------|----------------|------------------|
| Random Forest | 97.37%   | 100%          | 93%         | 96%           | 3                |
| AdaBoost      | **98.25%** | 100%        | 95%         | **98%**       | 2                |
| XGBoost       | 95.61%   | 100%          | 88%         | 94%           | 5                |

---

###  Key Observations:
- All models performed well with high accuracy and precision.
- **AdaBoostClassifier** showed the best overall performance with the highest accuracy and lowest number of false negatives.

###  Basic Error Analysis:

- **False Negatives**
  - Random Forest: 3
  - AdaBoost: 2
  - XGBoost: 5
- All models had **0 false positives**, which helps avoid unnecessary treatments or over-diagnosis.
- However, **false negatives are more dangerous** in medical scenarios, as they can result in missed cancer detection and delayed treatment.

#### Insight on XGBoost Errors:
From the classification metrics, we observed that XGBoost had the highest number of false negatives. These misclassifications likely occurred in **borderline cases**, where malignant samples had feature values (like radius_mean, area_mean, etc.) closer to benign thresholds — which made them harder to classify correctly.


### Final Conclusion:

- **AdaBoostClassifier** is the most suitable model for this problem:
  - It achieved the best accuracy and F1-score
  - It minimized false negatives, which is crucial in medical diagnostics

